src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+                    [=](bool result) {
+                        --_pendingAsyncJobs;
+                        if (!result) {
+                            processFileAnalyzeLocalInfo(item, path, localEntry, serverEntry, dbEntry, _queryServer);
+                        }
+                        QTimer::singleShot(0, _discoveryData, &DiscoveryPhase::scheduleMoreJobs);
+                    });
+                return;
+            }
+            // Turn new remote files into virtual files if the option is enabled.
+            if (_discoveryData->_syncOptions._newFilesAreVirtual && item->_type == ItemTypeFile) {
+                item->_type = ItemTypeVirtualFile;
+                item->_file.append(_discoveryData->_syncOptions._virtualFileSuffix);
+            }
+            processFileAnalyzeLocalInfo(item, path, localEntry, serverEntry, dbEntry, _queryServer);
+        };
+
+        if (!localEntry.isValid()) {
+            // Check for renames (if there is a file with the same file id)
+            bool done = false;
+            bool async = false;
+            // This function will be executed for every candidate
+            auto renameCandidateProcessing = [&](const OCC::SyncJournalFileRecord &base) {
+                if (done)
+                    return;
+                if (!base.isValid())
+                    return;
+
+                if (base._type == ItemTypeVirtualFileDownload) {
+                    // Remote rename of a virtual file we have locally scheduled
+                    // for download. We just consider this NEW but mark it for download.
+                    item->_type = ItemTypeVirtualFileDownload;
+                    done = true;
+                    return;
+                }
+
+                // Some things prohibit rename detection entirely.
+                // Since we don't do the same checks again in reconcile, we can't
+                // just skip the candidate, but have to give up completely.
+                if (base._type != item->_type && base._type != ItemTypeVirtualFile) {
+                    qCInfo(lcDisco, "file types different, not a rename");
+                    done = true;
+                    return;
+                }
+                if (!serverEntry.isDirectory && base._etag != serverEntry.etag) {
+                    /* File with different etag, don't do a rename, but download the file again */
+                    qCInfo(lcDisco, "file etag different, not a rename");
+                    done = true;
+                    return;
+                }
+
+                // Now we know there is a sane rename candidate.
+                QString originalPath = QString::fromUtf8(base._path);
+
+                // Rename of a virtual file
+                if (base._type == ItemTypeVirtualFile && item->_type == ItemTypeFile) {
+                    // Ignore if the base is a virtual files
+                    return;
```

Comment:
Shouldn't this prohibit renames on the server propagating to renaming the local virtual files? But since testRenameVirtual passes - why doesn't it?